### ISSUE TITLE
New addresses not being saved in the database

### ISSUE THREAD

**Issue by ashemsay** [spamgourmet/spamgourmet#13]
Hello,

I recently deployed an instance of spamgourmet using https://github.com/vasile-gh/spamgourmet-clone on Debian 10 / MariaDB.
Now I don't know if this issue is due to the deployment method or the source code, but if it's from the deployment script I'll open an issue there.

I noticed that remaining e-mails were not incremented in received messages Subjects when using a new address, I kept getting (<word>: message 1 of 3). So I went checking on the web interface and indeed i could not find said <word> in my addresses list so every time i received an e-mail for it, it was considered the first time that word was used.

I noticed the problem was coming from the insert query in `/usr/local/lib/spamgourmet/mailhandler/spameater` (l. 619) as it tried to insert a line with an empty `ExpireTime` and the database returned
```
ERROR 1366 (22007): Incorrect integer value: '' for column `sg`.`Emails`.`ExpireTime` at row 1
```
since this field is of type `BigInt` and `''` does not seem to be a valid `NULL` value.

I simply added this line above as workaround:

`$ExpireTime = 0 if $ExpireTime !~ /[0-9]+/;`

My question is: is this a bug or do you think a simple compatibility problem with my OS/Database or a deployment issue?

**Comment by coaxial** [comment]
This also seems to be happening on spamgourmet.com now: newly created addresses are never registered in the UI and the messages counter never goes up.

**Comment by stephankn** [comment]
I confirm this problem also occuring for me on spamgourmet.com.

I assume the database got stricter here and now refuses the value.

I wasn't aware, that it was even possible to have email adresses expire by date in addition to time by specifying a date like 2021-08-17 in the email.
This is parsed here:
https://github.com/spamgourmet/spamgourmet/blob/1c85c5a38222fecb1436cfce930486dbd89588f8/mailhandler/spameater#L346

Could it be, that the variable $ExpireTime should simply be initialized with a number instead of a string?
@ashemsay Can you please confirm?

https://github.com/spamgourmet/spamgourmet/blob/1c85c5a38222fecb1436cfce930486dbd89588f8/mailhandler/spameater#L257

to read

```
(0,'','','',0,0,0,0,0,0,0,'','',0,0,0,());
```

**Comment by gilch** [comment]
Related bbs.spamgourmet.com support thread: https://bbs.spamgourmet.com/viewtopic.php?t=1857

**Comment by gilch** [comment]
Has the fix been applied to spamgourmet.com yet?

**Comment by eegeeZA** [comment]
> Has the fix been applied to spamgourmet.com yet?
I don't believe so.

@josiah-hamilton will be able to comment on how deployments are done and if he has chosen to use this codebase.

**Comment by DaanDeJong2001** [comment]
@josiah-hamilton If you are running a slightly different code base, you can still do the one-line modification to initialize as a number, not string.

**Comment by josiah-hamilton** [comment]
Hey, ashemsay. I wish I had taken the wisdom from your trying to set this up on a newer instance. Thank you for identifying the issue and to stephankn's spelling out how to patch it here. I've committed it to the server codebase tonight and performed a couple rounds of two-way testing. Things seem to be working now. 
The running codebase is messy. I do not currently have a roadmap to simplify it or run off of this repo, but if we change too much it might get messy fast.

Thank you everybody for your hard work here. Please share if the issue as you see it is resolved.

**Comment by lwcorp** [comment]
Thank you! Can this ticket get closed?

**Comment by jgonggrijp** [comment]
> The running codebase is messy. I do not currently have a roadmap to simplify it or run off of this repo, but if we change too much it might get messy fast.

Maybe a practical solution would be to merge the running codebase into the current repository? Does the running codebase contain any code that cannot be shared with the public?

**Comment by josiah-hamilton** [comment]
Yes. The github codebase as it stands will work as a stand alone service barring some issues we encounter like this one, but I am informed by syskoll that there are some DDoS mitigation measures that were added in the early days (like before gmail came about and took everyone's attention away from us) and might be defeated with knowledge of how it works. These measures were not added to the sourceforge repo.

**Comment by jgonggrijp** [comment]
In that case, maybe you can do the opposite. If the current repo contains all code that can be public, and you keep a private branch that just adds the security measures like the DDoS mitigation, then you can repeatedly merge the former into the latter, as long as you don't go the opposite way. This should allow you to deploy from the public repo, if indirectly, so at least you can avoid an increasing mess.

**Comment by eegeeZA** [comment]
I think the idea of @jgonggrijp is a good one, but to instead make it a private fork. That way you can still merge the changes from this public repo into your private fork but keep pushing the changes to GitHub and not accidentally lose any code.

**Comment by mikedlr** [comment]
@josiah-hamilton do you think it would be reasonable to create a private fork as suggested by @eegeeZA but to give access to specific members of the spamgourmet project - specifically @eegeeZA and myself for now.  

This would enable us to compare the codebases and keep up to date as needed. This would also make it possible for us to see if we could provide some kind of simple modularisation / plugin type support which would mean that the standard codebase could support your private DDoS protection needs.  

(N.B. I haven't yet consulted @eegeeZA about this - please get explicit agreement from @eegeeZA that he can work on the private code base without sharing the secret information - I'm willing to do this also)

**Comment by vasile-gh** [comment]
1. Worth thinking about whether that code would be useful to SG clones?
2. If yes, it would be good if you could think of a mechanism to (perhaps) make that code "privately" available to people running clones - I understand there's a risk there so maybe only share it with people who prove that they run an operational SG clone?

**Comment by mikedlr** [comment]
@vasile-gh obviously DDoS resistance is useful for the people running clones, so yes. I think it would be useful in principle
however there's also been quite a bit of advance in DDoS projection for mail generally.  What we'd probably do is look at it and see if we can find a way to add other more standard protections that, since they are already public, won't need as much secrecy.

**Comment by eegeeZA** [comment]
You have my agreement @mikedlr to keep the code private. I don't mind signing an NDA if that's required. Also like the idea of checking to see if the DDoS code can be swapped out for a modern option.

**Comment by josiah-hamilton** [comment]
Hey everybody,

I've sent this thread to Syskoll. We opted to keep it off of CVS under his and my dad's advisement. It's not like an NDA is remotely what's needed because we're not licensing this from some third party vendor. 
A couple weeks ago I started spinning up a gitea instance at home to manage the private repo before I stalled out on that when discovering a bunch of stuff my dad had added around middle of 2019 (I presume to forward some of his disposables to my mom for e.g. billing before his brain damage kicked into high gear, based on the table entries). AlternativeRecipientAddress table, and it works! Just no web interface for it. That feature is a non-sequitur in this discussion, but it will teach me how to do branching and PRs in git in order to work it into master. 

Topic at hand, I am not closed to the idea of newer DDoS mitigations that would work on single host instances. One that we could proudly display rather than kind of shamefully acknowledge that understanding its operation makes it easy to defeat. Also I have to locate it in the messy and scattered directories. My dad's test-in-prod is organized exactly how you'd expect it to.  But first I want to talk it out with Syskoll (I'm not sure if he's publicly disclosed his name)

**Comment by SysKoll** [comment]
Hey folks,

I am also in favor of a private fork where we can keep the more sensitive code.

I truly despise the notion of security through obscurity, but in this case, I am reluctantly forced to admit it's useful. Historically, spamgourmet was constantly brought down by malicious DDoS until Josh and I implemented these measures. Now that spam mitigation is prevalent among email service suppliers, spamgourmet is probably not in the gunsight of spammers anymore, but it is such a pain to remediate even one successful DDoS that I still want some degree of obscurity for the anti-DDoS code.

So bottom line, count me in for a private repo containing a superset of the public repo.

We'll need to define procedures for maintaining the two repo (or forks) but I believe that the private repo should simply accept all the changes from the public repo (which would be the upstream). Changes for the sensitive code will never leave the private repo.

Comments? Thoughts?

**Comment by jgonggrijp** [comment]
Hey @SysKoll, nice to have you on this party as well.

I have nothing new to add, but I agree with earlier comments by @vasile-gh and @mikedlr, on the counts that there should be a way for SG forks (or at least proven secondary SG installations) to use the secret security fixes and that it is probably worthwhile to look out for DDoS mitigations that still work when known to the attacker. Until that time, though, I'm not opposed to having this secondary private repo.
